### PR TITLE
Compact Primitive/String and Compact Arrays [API-1297]

### PR DESCRIPTION
**Pull request by sancar**

### CODE REVIEW

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h [hunk 1]
```diff
@@ -148,13 +167,120 @@ compact_reader::read()
...
+    int32_t data_length = object_data_input.read<int32_t>();
+    int32_t item_count = object_data_input.read<int32_t>();
+    int data_start_pos = object_data_input.position();
+    auto offset_reader = get_offset_reader(data_length);
+    int offsets_position = data_start_pos + data_length;
+    for (int i = 0; i < item_count; ++i) {
+        int offset = offset_reader(object_data_input, offsets_position, i);
+        if (offset == util::Bits::NULL_ARRAY) {
+            BOOST_THROW_EXCEPTION(exception_unexpected_null_value_in_array(
+              field_name, method_suffix));
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h:261]: There is no field name on field descriptor in c++

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact.h [hunk 2]
```diff
@@ -274,15 +374,60 @@ class HAZELCAST_API compact_reader
...
+      const pimpl::field_descriptor& field_descriptor);
+    template<typename T>
+    boost::optional<T> get_nullable_array_as_primitive_array(
+      const pimpl::field_descriptor& field_descriptor,
+      const std::string& field_name,
+      const std::string& method_suffix);
+    static std::function<
+      int32_t(serialization::object_data_input&, uint32_t, uint32_t)>
+    get_offset_reader(int32_t data_length);
+    exception::hazelcast_serialization exception_unexpected_null_value_in_array(
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.h:422]: static methods can not be const.

[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.h:422]: I will make the names aligned though.

[the remark this answers is not in the document]
Comment (srknzl) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.h:422]: `exception_unexpected_null_value_in_array` is not static at the moment, right?

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h [hunk 3]
```diff
@@ -148,13 +167,120 @@ compact_reader::read()
...
+    } else if (field_descriptor.field_kind == nullable_field_kind) {
+        return get_nullable_array_as_primitive_array<T>(
+          field_descriptor, field_name, method_suffix);
+    }
+    throw unexpected_field_kind(field_descriptor.field_kind, field_name);
+}
+
+template<typename T>
+boost::optional<std::vector<boost::optional<T>>>
+compact_reader::get_array_of_variable_size(
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h:216]: Converting all to read.

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h [hunk 4]
```diff
@@ -148,13 +167,120 @@ compact_reader::read()
...
+  const pimpl::field_descriptor& field_descriptor)
+{
+    int32_t position = read_var_size_position(field_descriptor);
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h:219]: I was fixing this one in upcoming pr's but now moving it to here.

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.h:422]: Yes.

**Review by srknzl — APPROVED** [review]

File: hazelcast/src/hazelcast/client/compact.cpp [hunk 5]
```diff
@@ -199,8 +300,35 @@ compact_reader::get_field_descriptor(const std::string& field_name,
     return field_descriptor;
 }
 
+std::function<int32_t(serialization::object_data_input&, uint32_t, uint32_t)>
+compact_reader::get_offset_reader(int32_t data_length)
+{
+    if (data_length < pimpl::offset_reader::BYTE_OFFSET_READER_RANGE) {
+        return pimpl::offset_reader::get_offset<int8_t>;
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/compact.cpp:307]: Can you change these to singleton usages as it is at Java?

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h [hunk 6]
```diff
@@ -148,11 +168,125 @@ compact_reader::read()
...
+
+template<typename T>
+boost::optional<T>
+compact_reader::read_array_of_primitive(
+  const std::string& field_name,
+  enum pimpl::field_kind field_kind,
+  enum pimpl::field_kind nullable_field_kind,
+  const std::string& method_suffix)
+{
+    auto& field_descriptor = get_field_descriptor(field_name);
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h:204]: Methodname is consistent with the class name in Java. I think it makes more sense to fix at java. 
Noted here to fix it. https://hazelcast.atlassian.net/browse/API-1320

**Pull request merged**